### PR TITLE
new game: memory

### PR DESCRIPTION
**Pull request by privat**
The code is good enough and the game complete enough I think.

### CODE REVIEW

**Comment by privat** [comment]
The android apk is available at http://info.uqam.ca/~privat/memory.apk

File: contrib/memory/README.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# Memorize Shapes and Colors
+
+A memory-based game where figures are cliqued in sequence by the computer and should be replayed by the player in the same order.
+As the player progress, more figure are added and the sequences to remember become longer.
```
Comment (R4PaSs) [inline contrib/memory/README.md]: :s/progress/progresses

Comment (R4PaSs) [inline contrib/memory/README.md]: :s/figure/figures

File: contrib/memory/README.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+The player can make up to 2 errors to solve a single level.
+At the 3rd error, the level has to be replayed.
+
+The game use a very simple user interface and features big figures with bright colors and simple distinguishable shapes; that makes it suitable for young children.
+
+The game offers three mode (difficulty level)
```
Comment (R4PaSs) [inline contrib/memory/README.md]: :s/mode/modes

File: contrib/memory/README.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+Easy (for young children):
+
+* Start with 2 figures, 1 to remember
+* Figure are easily distinguishable
```
Comment (R4PaSs) [inline contrib/memory/README.md]: :s/figure/figures

File: contrib/memory/org.nitlanguage.memory.txt [hunk 4]
```diff
@@ -0,0 +1,16 @@
+Categories:Nit,Games
```
Comment (R4PaSs) [inline contrib/memory/org.nitlanguage.memory.txt]: Not really a remark, but I wonder what this file is for, is it some kind of Android metadata ?

File: contrib/memory/src/memory.nit [hunk 5]
```diff
@@ -0,0 +1,994 @@
...
+# * [X] Saved high scores
+# * [ ] Level selection
+#
+# The remaining issues are
+#
+# * Crappy event system
+# * Crappy UI element placement
+module memory is
+	app_name("Memorize Shapes and Colors")
+        app_version(0, 1, git_revision)
```
Comment (R4PaSs) [inline contrib/memory/src/memory.nit]: Indentation is weird

File: contrib/memory/org.nitlanguage.memory.txt (hunk 4 above)
Comment (privat) [inline contrib/memory/org.nitlanguage.memory.txt]: It's for fdroid

**Comment by privat** [comment]
reroll: spelling by @R4PaSs

**Comment by R4PaSs** [comment]
:shipit: then !

File: contrib/memory/src/memory.nit [hunk 6]
```diff
@@ -0,0 +1,994 @@
...
+				end
+				return true
+			end
+			print c
+
+			if is_menu then
+				return false
+			end
+
+			if c == "[+]" or c == "="then
```
Comment (xymus) [inline contrib/memory/src/memory.nit]: What is `[+]`?

File: contrib/memory/src/memory.nit [hunk 7]
```diff
@@ -0,0 +1,994 @@
...
+			if c == "4" or c == "escape" then
+				if is_menu then
+					# quit = true # broken
+					new_menu
+				else
+					new_menu
+				end
+				return true
+			end
+			print c
```
Comment (xymus) [inline contrib/memory/src/memory.nit]: If this is a debug print you should add some context to the print for it to be helpful later.

File: contrib/memory/src/memory.nit (hunk 6 above)
Comment (privat) [inline contrib/memory/src/memory.nit]: it seems to be numpad +, it is what the undocumented print was used for

File: contrib/memory/src/memory.nit [hunk 8]
```diff
@@ -0,0 +1,994 @@
...
+			var path = new_path(from, to, 5.0)
+			b.path = path
+			b.ttl = 5.0
+			b.hard = false
+		end
+		ttl = 6.0
+	end
+
+	# Shuffle the buttons so that each on moves.
+	fun derangement
```
Comment (xymus) [inline contrib/memory/src/memory.nit]: `derangement & shuffle`... It looks like you missed inspiration at this point, and the docs are almost the same. Why not `shuffle_core` or `shuffle_shared` to differentiate it from its users?

Comment (privat) [inline contrib/memory/src/memory.nit]: `derangement` is the algorithmic concept: https://en.wikipedia.org/wiki/Derangement

shuffle is what is happening on the screen. But I can improve the doc

File: lib/core/math.nit [hunk 9]
```diff
@@ -333,6 +333,45 @@ redef class Float
...
+	# assert  1.00.qerp(0.0, 32.0, 128.0) == 128.0
+	# ~~~
+	fun qerp(a, handle, b: Float): Float do
+		var p = self
+		var i = 1.0 - p
+		var r =     i*i * a +
+			2.0*i*p * handle +
+			    p*p * b
+		return r
+	end
```
Comment (xymus) [inline lib/core/math.nit]: Good idea moving it to math, and I love the code art!

**Comment by xymus** [comment]
+1

File: contrib/memory/src/memory.nit (hunk 8 above)
Comment (xymus) [inline contrib/memory/src/memory.nit]: Good to know!

**Comment by privat** [comment]
last(?) reroll, for @xymus

**Pull request merged**